### PR TITLE
Hide the implementation of the transport layer

### PR DESCRIPTION
**Pull request by riemass**
Thir PR hides the implementation details of `octet_stream::transport` and `ssl::transport`. The policy is pushed to the heap and the constructor parameters have been modified accordingly. 
Relates #1254.

### CODE REVIEW

**Comment by riemass** [comment]
@Neverlord I was unsure whether to remove the in-place object optimization. Probably something like this would be acceptable: 
```
union {
  policy_impl default_policy_;
  std::unique_ptr<policy> user_policy_;
};
policy* policy_ = ...; 
```

File: libcaf_net/caf/net/octet_stream/transport.cpp [hunk 1]
```diff
@@ -12,361 +12,478 @@
...
+  byte_buffer write_buf_;
+
+  /// Processes incoming data and generates outgoing data.
+  upper_layer_ptr up_;
+
+  /// Points to the associated socket manager below.
+  socket_manager* parent_ = nullptr;
+
+  /// Configures how we read and write to the socket.
+  std::unique_ptr<policy> policy_ = nullptr;
```
Comment (Neverlord) [inline libcaf_net/caf/net/octet_stream/transport.cpp:468]: That's just what the default constructor will do anyways.

```suggestion
  std::unique_ptr<policy> policy_;
```

File: libcaf_net/caf/net/octet_stream/transport.hpp [hunk 2]
```diff
@@ -30,194 +27,32 @@ class CAF_NET_EXPORT transport : public socket_event_layer,
...
-  /// Stores what the user has configured as read threshold.
-  uint32_t min_read_size_ = 0;
-
-  /// Stores what the user has configured as max. number of bytes to receive.
-  uint32_t max_read_size_ = 0;
-
-  /// Stores how many bytes are currently buffered, i.e., how many bytes from
-  /// `read_buf_` are filled with actual data.
-  size_t buffered_ = 0;
+  virtual byte_buffer& read_buffer() noexcept = 0;
```
Comment (Neverlord) [inline libcaf_net/caf/net/octet_stream/transport.hpp:39]: I couldn't find references to this member function. Do we really need `read_buffer()` and `write_buffer()`?

File: libcaf_net/caf/net/octet_stream/transport.hpp [hunk 3]
```diff
@@ -30,194 +27,32 @@ class CAF_NET_EXPORT transport : public socket_event_layer,
...
-  byte_buffer read_buf_;
+  virtual byte_buffer& write_buffer() noexcept = 0;
 
-  /// Caches outgoing data.
-  byte_buffer write_buf_;
+  virtual const byte_buffer& write_buffer() const noexcept = 0;
 
-  /// Processes incoming data and generates outgoing data.
-  upper_layer_ptr up_;
+  virtual class upper_layer& upper_layer() noexcept = 0;
```
Comment (Neverlord) [inline libcaf_net/caf/net/octet_stream/transport.hpp:47]: Do we really need this getter?

**Review by Neverlord — CHANGES_REQUESTED** [review]

File: libcaf_net/caf/net/octet_stream/transport.hpp (hunk 2 above)
Comment (riemass) [reply libcaf_net/caf/net/octet_stream/transport.hpp]: Good catch.

**Pull request merged**